library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/recorder/SnapshotProducer.kt

```diff
@@ -6,13 +6,305 @@
 
 package com.datadog.android.sessionreplay.recorder
 
+import android.graphics.Typeface
+import android.graphics.drawable.ColorDrawable
+import android.graphics.drawable.Drawable
+import android.graphics.drawable.InsetDrawable
+import android.graphics.drawable.RippleDrawable
+import android.os.Build
+import android.view.Gravity
 import android.view.View
+import android.view.ViewGroup
+import android.view.ViewStub
+import android.widget.Button
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.appcompat.widget.ActionBarContextView
+import androidx.appcompat.widget.Toolbar
+import com.datadog.android.sessionreplay.model.MobileSegment
+import java.util.LinkedList
+import java.util.Locale
 
-@Suppress("UNUSED_PARAMETER")
 internal class SnapshotProducer {
 
-    fun produce(rootView: View, scale: Float): Node? {
-        // TODO: RUMM-2356
-        return null
+    private val systemViewIds by lazy {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES
+            .LOLLIPOP
+        ) {
+            setOf(android.R.id.navigationBarBackground, android.R.id.statusBarBackground)
+        } else {
+            emptySet()
+        }
+    }
+
+    fun produce(rootView: View, pixelsDensity: Float): Node? {
+        return rootView.toNode(pixelsDensity)
+    }
+
+    @Suppress("ComplexMethod")
+    private fun View.toNode(pixelsDensity: Float): Node? {
+        if (isNotValid()) {
```

Comment:
Same here, using extension makes things ambiguous here. Does `isNotValid` refer to the implicit `this: View`, or to the parent `this: SnapshotProducer`?